### PR TITLE
Handle the error when SCIM request fails in Outbound provisioning

### PR DESCRIPTION
**Pull request by kayathiri4**
Purpose
Fixes the issue:
If the SCIM request fails by any other listener or even from the userstore level, the user will be outbound provisioned.

Solution:
Introduce a error listener to revert the provisioning if SCIM request fails.

Related issue: https://github.com/wso2/product-is/issues/12074

### CODE REVIEW

File: components/provisioning/org.wso2.carbon.identity.provisioning/src/main/java/org/wso2/carbon/identity/provisioning/listener/OutboundProvisioningErrorListener.java [hunk 1]
```diff
@@ -0,0 +1,463 @@
...
+        if (log.isDebugEnabled()) {
+            log.debug("Adding domain name : " + domainName + " to user : " + userName);
+        }
+        String domainAwareName = UserCoreUtil.addDomainToName(userName, domainName);
+
+        ProvisioningEntity provisioningEntity = new ProvisioningEntity(
+                ProvisioningEntityType.USER, domainAwareName, ProvisioningOperation.PUT,
+                outboundAttributes);
+
+        // set the in-bound attribute list.
```
[the remark this answers is not in the document]
Comment (somindatommy) [reply components/provisioning/org.wso2.carbon.identity.provisioning/src/main/java/org/wso2/carbon/identity/provisioning/listener/OutboundProvisioningErrorListener.java:300]: Fix in all places

File: components/provisioning/org.wso2.carbon.identity.provisioning/src/main/java/org/wso2/carbon/identity/provisioning/listener/OutboundProvisioningErrorListener.java [hunk 2]
```diff
@@ -0,0 +1,463 @@
...
+                            IdentityProvisioningConstants.USERNAME_CLAIM_URI, null, null, false),
+                    Arrays.asList(new String[]{userName}));
+        }
+
+        String domainName = UserCoreUtil.getDomainName(userStoreManager.getRealmConfiguration());
+        if (log.isDebugEnabled()) {
+            log.debug("Adding domain name : " + domainName + " to user : " + userName);
+        }
+        String domainAwareName = UserCoreUtil.addDomainToName(userName, domainName);
+
```
Comment (somindatommy) [inline components/provisioning/org.wso2.carbon.identity.provisioning/src/main/java/org/wso2/carbon/identity/provisioning/listener/OutboundProvisioningErrorListener.java:295]: Remove unnecessary newlines. Fix in all places

File: components/provisioning/org.wso2.carbon.identity.provisioning/src/main/java/org/wso2/carbon/identity/provisioning/listener/OutboundProvisioningErrorListener.java [hunk 3]
```diff
@@ -0,0 +1,463 @@
...
+    }
+
+    /**
+     * Method to delete a role.
+     * @param roleName - Name of the role
+     * @param userStoreManager - User Store
+     * @return boolean
+     * @throws UserStoreException
+     */
+    private boolean deleteRole(String roleName, org.wso2.carbon.user.api.Permission[] permissions,
```
Comment (somindatommy) [inline components/provisioning/org.wso2.carbon.identity.provisioning/src/main/java/org/wso2/carbon/identity/provisioning/listener/OutboundProvisioningErrorListener.java:345]: Fix the comment

File: components/provisioning/org.wso2.carbon.identity.provisioning/src/main/java/org/wso2/carbon/identity/provisioning/listener/OutboundProvisioningErrorListener.java [hunk 4]
```diff
@@ -0,0 +1,463 @@
...
+        return true;
+    }
+
+    /**
+     * Method to get the priority of the listener.
+     * @return {int} priority
+     */
+    public int getExecutionOrderId() {
+        return 9;
+    }
```
Comment (somindatommy) [inline components/provisioning/org.wso2.carbon.identity.provisioning/src/main/java/org/wso2/carbon/identity/provisioning/listener/OutboundProvisioningErrorListener.java:462]: Fix formatting and method comment

File: components/provisioning/org.wso2.carbon.identity.provisioning/src/main/java/org/wso2/carbon/identity/provisioning/internal/ProvisioningServiceDataHolder.java [hunk 5]
```diff
@@ -84,6 +86,16 @@ public void setEntitlementService(EntitlementService entitlementService) {
...
     }
+
+    public void setRolePermissionManagementService(RolePermissionManagementService rolePermissionManagementService) {
+
+        this.rolePermissionManagementService = rolePermissionManagementService;
+    }
+
+    public RolePermissionManagementService getRolePermissionManagementService() {
+
+        return rolePermissionManagementService;
```
[the remark this answers is not in the document]
Comment (kayathiri4) [reply components/provisioning/org.wso2.carbon.identity.provisioning/src/main/java/org/wso2/carbon/identity/provisioning/internal/ProvisioningServiceDataHolder.java:105]: Addressed with dd48ca7fb7eab31e8e4dd2b8890c8be1e21002c9

File: components/provisioning/org.wso2.carbon.identity.provisioning/src/main/java/org/wso2/carbon/identity/provisioning/listener/OutboundProvisioningErrorListener.java [hunk 6]
```diff
@@ -0,0 +1,463 @@
...
+                String[] permissionList = ProvisioningServiceDataHolder.getInstance().getRolePermissionManagementService()
+                        .getRolePermissions(roleName, tenantId);
+                if (permissionList.length > 0) {
+                    for (int index=0; index<permissionList.length; index++) {
+                        permissions[index] = new org.wso2.carbon.user.api.Permission(permissionList[index],
+                                UserMgtConstants.EXECUTE_ACTION);
+                    }
+                }
+            } catch (Exception e) {
+                // Do nothing
```
[the remark this answers is not in the document]
Comment (kayathiri4) [reply components/provisioning/org.wso2.carbon.identity.provisioning/src/main/java/org/wso2/carbon/identity/provisioning/listener/OutboundProvisioningErrorListener.java]: The exception is thrown while getting the permissions for particular role. Handled it by adding debug logs.

**Review by ShanChathusanda93 — APPROVED** [review]

**Pull request merged**